### PR TITLE
Integration test for the WSL OOBE setup wizard

### PR DESCRIPTION
**Pull request by jpnurmi**
A full end-to-end integration test that uses Flutter's integration test driver to run the app, navigate through the wizard, and enter test values. Finally, it verifies that the appropriate values have been written to `.subiquity/etc/wsl.conf` and `.subiquity/etc/ubuntu-wsl.conf`.

Test cases:
- Basic setup: _Select language_ and fill up _Profile setup_ 
  - TODO: is the profile information stored anywhere in dry-run mode?
- Advanced setup: select _Show advanced options_ and fill up _Advanced setup_
- Reconfiguration: pass `--reconfigure` and fill up _Configuration UI_

To run the test outside of IDE, run `flutter test integration_test/` in `packages/ubuntu_wsl_setup`.

### CODE REVIEW

**Comment by jpnurmi** [comment]
https://user-images.githubusercontent.com/140617/136602013-b968d96c-a7ac-405e-b82f-490ea0536919.mp4

File: packages/ubuntu_wsl_setup/integration_test/ubuntu_wsl_setup_test.dart [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+      await tester.pump();
+    }
+    await tester.pumpAndSettle();
+
+    await tester.tapContinue();
+  }
+
+  Future<void> testProfileSetup(
+    WidgetTester tester,
+    IdentityData? profile, {
```
Comment (didrocks) [inline packages/ubuntu_wsl_setup/integration_test/ubuntu_wsl_setup_test.dart:50]: I didn’t see any code path where you call it with null, is it intended to be nullable?
Same for the 3 other fields

File: packages/ubuntu_wsl_setup/integration_test/ubuntu_wsl_setup_test.dart [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+          equals(config.motdWSLnewsenabled.toString()));
+    });
+  }
+
+  testWidgets('basic setup', (tester) async {
+    app.main(<String>[]);
+    await waitForSubiquityServer();
+    await tester.pumpAndSettle();
+
+    await testSelectYourLanguage(tester, language: 'Français');
```
Comment (didrocks) [inline packages/ubuntu_wsl_setup/integration_test/ubuntu_wsl_setup_test.dart:233]: Merci ! ;)

File: packages/ubuntu_wsl_setup/integration_test/ubuntu_wsl_setup_test.dart [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+
+    await verifyBaseConfig(config);
+  });
+
+  testWidgets('reconfiguration', (tester) async {
+    app.main(<String>['--reconfigure']);
+    await waitForSubiquityServer();
+    await tester.pumpAndSettle();
+
+    await testAdvancedSetup(tester);
```
Comment (didrocks) [inline packages/ubuntu_wsl_setup/integration_test/ubuntu_wsl_setup_test.dart:298]: Mind adding a comment that this is testing the first screen (WSLConfigurationBase)?

File: packages/ubuntu_wsl_setup/integration_test/ubuntu_wsl_setup_test.dart [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+      '.subiquity',
+      'etc',
+      fileName,
+    );
+    final lines = File(path).readAsLinesSync();
+    return test(Config.fromStrings(lines));
+  }
+
+  Future<void> verifyBaseConfig(WSLConfigurationBase config) async {
+    await verifyIniFile('wsl.conf', (ini) {
```
Comment (didrocks) [inline packages/ubuntu_wsl_setup/integration_test/ubuntu_wsl_setup_test.dart:198]: Note: we are more used to use golden files as best practice where you can pass scenarios (base inputs) and serialize the results in a file, then comparing that file (with a mecanism to refresh them with running the tests with -update). Not a blocker as this is also valid (a little bit less agile), but maybe something to keep in mind? I’m happy to expand my explanations if needed

**Review by didrocks — COMMENTED** [review]
> TODO: is the profile information stored anywhere in dry-run mode?
Indeed, this is a task that is on Patrick’s roadmap. Please have a chat with him (I think writing to a file in .subiquity for both actions would be good, or adding a mock adduser command which does this writing)

Some questions & small remarks attached above

**Comment by jpnurmi** [comment]
Hmm, sometimes the .conf files don't exist. I'm not sure if it's because the GUI is abruptly killing the subiquity process or what. Needs investigation...

**Comment by jpnurmi** [comment]
Rebased, cleaned up, and squashed. Depends on https://github.com/canonical/ubuntu-desktop-installer/pull/504.

File: packages/ubuntu_wsl_setup/integration_test/ubuntu_wsl_setup_test.dart (hunk 1 above)
Comment (jpnurmi) [reply packages/ubuntu_wsl_setup/integration_test/ubuntu_wsl_setup_test.dart]: The idea was that one could surf through pages without having to pass anything in test cases where that page is irrelevant.

File: packages/ubuntu_wsl_setup/integration_test/ubuntu_wsl_setup_test.dart (hunk 4 above)
Comment (jpnurmi) [reply packages/ubuntu_wsl_setup/integration_test/ubuntu_wsl_setup_test.dart]: It turns out Flutter exposes the information whether `--update-goldens` was passed, so we can use the "standard" `flutter-test --update-goldens integration_test`, after all. There are instructions in the [README](https://github.com/canonical/ubuntu-desktop-installer/blob/7e4f79fe0ad27aae12439b9244f51c2f9e5f28ca/packages/ubuntu_wsl_setup/integration_test/README.md#update-golden-files).

**Comment by didrocks** [comment]
LGTM, the golden file is testing the very important use-case (what should I say, even the MAIN and only one we all care, right? :p)

I wonder if we can have an integration test doing no change and checking that no change is done on the wsl.conf. We can have with a pre-existing wsl.conf, and one without one. What do you think?

**Comment by jpnurmi** [comment]
> LGTM, the golden file is testing the very important use-case (what should I say, even the MAIN and only one we all care, right? :p)

Yeah, it's great especially with the update flag when needed.

> I wonder if we can have an integration test doing no change and checking that no change is done on the wsl.conf. We can have with a pre-existing wsl.conf, and one without one. What do you think?

Sounds good! :+1:

**Review by didrocks — APPROVED** [review]

**Pull request merged**